### ISSUE TITLE
float's inf is Inf on SPARC Solaris 11 

### ISSUE THREAD

**Issue by dimpase** [gap-system/gap#2193]
### Observed behaviour
```
########> Diff in /datapool/dima/Sage/gap/tst/testinstall/float.tst:130
# Input is:
posinf := Float(infinity);
# Expected output:
inf
# But found:
Inf
```
---and similar messages in all tests involving `inf` in `testinstall/float.tst`

### Expected behaviour
tests should pass

```
*********   GAP 4.8.8-5085-gb6d4e8e-dirty of today
 *  GAP  *   https://www.gap-system.org
 *********   Architecture: sparc-sun-solaris2.11-default64
 Configuration:  gmp 6.1.2, KernelDebug
 Loading the library and packages ...
 Packages:   AClib 1.2, Alnuth 3.1.0, AtlasRep 1.5.1, AutPGrp 1.8, CRISP 1.4.4, Cryst 4.1.13, CrystCat 1.1.6, CTblLib 1.2.2, FactInt 1.6.1, FGA 1.3.1, 
             GAPDoc 1.6.1, IRREDSOL 1.4, LAGUNA 3.8.0, Polenta 1.3.8, Polycyclic 2.11, PrimGrp 3.3.0, RadiRoot 2.7, ResClasses 4.7.1, smallgrp 1.2, 
             Sophus 1.23, SpinSym 1.5, TomLib 1.2.6, TransGrp 2.0.2, utils 0.53
```
This is git master of 19.02.2018. Same errors observed on GAP 4.9 beta of 1st of Feb,
too. Built with gcc 5.4.0.

**Comment by dimpase** [comment]
This can be traced to
```
gap> MACFLOAT_STRING("inf");
Inf
```
That is, either `MACFLOAT_STRING` capitalises `i`, or an output routine does it.

**Comment by ChrisJefferson** [comment]
We just use snprintf to print floating point numbers. The C standard says that 'inf' should be printed. What does the following C program print?

```
#include <stdio.h>

int main(void)
{
    double val = 10;
    for(int i = 0; i < 10000; ++i) {
        val = val * val;
    }
    printf("%g\n", val);
}
```

**Comment by fingolfin** [comment]
@ChrisJefferson where did you get that from? In POSIX, I read this (on <http://pubs.opengroup.org/onlinepubs/009696799/functions/printf.html>):
> A double argument representing an infinity shall be converted in one of the styles "[-]inf" or "[-]infinity" ; which style is implementation-defined.
Of course `Inf` still would not conform to POSIX, but it seems we can't rely on `inf` being the output either. I guess one possible workaround here would be to try and detect +/-inf values and print a custom string for them. This might be necessary on certain POSIX compliant systems, too.

BTW, is `NaN` printed "correctly" on Solaris?

**Comment by dimpase** [comment]
using `L` modifier in format results in printing `1.19446e-4946`, both for `inf` and `NaN`...

**Comment by dimpase** [comment]
I have also checked that calling `tolower()` on the 1st char of `res` in `sprintf(res,"%g\n",HUGE_VALUE)` duly makes it `inf` (and it really is sprinted `Inf`, no funky locale quirks here).

**Comment by dimpase** [comment]
there are no tests for `NaN`. I suppose there should be a couple. Anyway, it's `NaN` on Solaris and `nan` on my Linux laptop, YMMV, I guess. Should we always print `NaN` as `nan`, dropping sign?

**Comment by fingolfin** [comment]
Yes, we should always print `nan` resp. `-nan`. Also, the sign should be determined using `signbit(val)` and not `val > 0`. And of course there should be a brief comment explaining what is going on. And finally, `FuncSTRING_DIGITS_MACFLOAT` also needs a similar change -- I suspect the best route here is to move the `snprintf` into a new helper function  `PrintMacfloatToBuf(char *buf, size_t bufsize, Double val)` and then use that helper in both `PrintMacfloat` and `FuncSTRING_DIGITS_MACFLOAT`. (And while I am looking at it: `TOPRINTFFORMAT` can and should be removed)

**Comment by fingolfin** [comment]
Hmm, it is weird that we always print negative nan as `nan`, too. I thought POSIX required it to be printed as `-nan`? 

Anyway, I just noticed that our sign handling for floats has serious issues (I discussed this before with @laurentbartholdi). I will submit an issue on that.

**Comment by dimpase** [comment]
IMHO the sign of `nan` is not defined in the standards; it's machine-dependent (thus). In fact something like 
```
gap> Float(infinity)-Float(infinity);
```
prints `-nan` on a Linux box and `NaN` on Solaris.
Thus I think it's a good idea to always print `nan`.

**Comment by dimpase** [comment]
On the other hand I see no harm in `val>0` for `val` being an `inf`. It's in accordance with IEEE floats standards that this is possible.

**Comment by fingolfin** [comment]
IMHO? Perhaps you mean AFAIK? As this is not a matter of opinion :-).

I am not sure if printing all nan as positive is a good solution then. Because `SignBit` and `SignFloat` would then still print differing results, which would show up in tests (e.g. in the tests I am writing right now, as I am fixing the C version of `SignFloat`, which is still using the incorrect tests `val > 0.0` to determine the sign).

**Comment by fingolfin** [comment]
The relations `>`, `=` and `<` are *not* mutually exclusive for IEEE floats, i.e. don't satisfy trichotomy, hence code determining sign has to be very careful. Sure, it is OK for inf, but I would still strongly suggest to *never* use `val > 0.0` (and certainly not `val > 0`) simply to avoid the habit in places where it *does* matter. GAP currently *still* uses it in various places (despite going over this already once before with @laurentbartholdi ) and as a result, produces inconsistent or even invalid results in certain cases.

**Comment by fingolfin** [comment]
I think @dimpase is right and signs of nan are implementation specific. So we could print them always as `nan`, and also adjust the sign computations to always indicate a positive sign for all nan. Or, we don't care about the latter and just don't add in tests for the sign of nan.

**Comment by dimpase** [comment]
tests for nan are useful just for the sake of checking the underlying
implementation.

**Comment by fingolfin** [comment]
Of course tests for nan are useful, but not necessarily for the *sign* of nans.

**Comment by fingolfin** [comment]
Another (somewhat hypothetical) problem with NaN: depending on the system and compiler flags, we could encounter signalling NaNs and these could trap, i.e. result in GAP crashing. We can avoid this by using `isless` etc. from C99, to perform comparisons on the C level; or by explicitly catching NaNs and treating them specially (of course for signs, but also for other things). 
But I am not sure one actually ever gets signalling NaNs unless one uses special compiler options and/or uses non-portable APIs to enable them, so I am probably worrying for no good reason...

... but man, I really wish I didn't have to deal with this floating point shh..tuff.

**Comment by dimpase** [comment]
Nothing should prevent GAP from producing an error on an attempt to get the
sign of NaN.

Regarding the comparisons, I agree that == must never be used on floats. Do
you mean to say that instead of val>0 one should do val>eps, for certain
eps dependent on hardware?

**Comment by dimpase** [comment]
Are we still proceeding with changing `PrintMacfloat` as proposed?

**Comment by dimpase** [comment]
Oracle support told us that using Oracle-patched gcc-5.4.0 (shipped with Solaris 11) with the switch `-std=c99` (or `c1x`) results in getting `inf` and `nan` printed. The further said that the "default" is `-std=c90`, which is non sequitur, as it does compile code which is not `c90` without any switches.

I'm planning to dig through their multi-gigabyte 
  http://www.oracle.com/technetwork/opensource/systems-solaris-1562786.html to dig up the patches needed and perhaps try to offer them to upstream.
(there are also some seemingly relevant github repos by Oracle, such as https://github.com/oracle/solaris-userland/tree/master/components/gcc7 although they look more experimental stuff)

**Comment by dimpase** [comment]
They've gave details there. At least it's clear what's going on, more or less.

**Comment by dimpase** [comment]
This is fixed in [gcc 8.1](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=40411
)  (sic!)
